firewood/src/merkle/proof.rs

```diff
@@ -473,41 +480,52 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
 
         match new_node.inner() {
             NodeType::Leaf(n) => {
-                let cur_key = n.path().0.as_ref();
+                let paths_match = n.path.iter().copied().eq(&mut key_nibbles);
 
-                if !key.contains_other(cur_key) {
-                    return Ok((addr, None, 0));
+                if !paths_match {
+                    return Ok((addr, None, Nibbles::new(&[]).into_iter().peekable()));
                 }
 
                 let subproof = SubProof::Data(n.data().to_vec());
 
-                Ok((addr, subproof.into(), cur_key.len()))
+                Ok((addr, subproof.into(), key_nibbles))
             }
             NodeType::Extension(n) => {
-                let cur_key = n.path.0.as_ref();
-
-                if !key.contains_other(cur_key) {
-                    return Ok((addr, None, 0));
+                let paths_match = n
+                    .path
+                    .iter()
+                    .copied()
+                    .all(|nibble| Some(nibble) == key_nibbles.next());
+
+                if !paths_match {
+                    return Ok((addr, None, Nibbles::new(&[]).into_iter().peekable()));
```

Comment:
question (please answer by adding a comment if possible)

Why are we returning an empty nibbles iterator instead of None or something like that?

Reply:
I'll add doc-comments to the function. 

It's better to handle iteration generically. An empty iterator means there are no more nibbles to iterate over and we should handle the result accordingly.

Reply:
Doc comments added. I'm going to resolve this comment as you can open new comments on the new doc-comments directly